### PR TITLE
Improve rustdoc json tests

### LINKED ISSUE #100588: JsonDocCk: Automaticly infer file name
Right now in jsondocck all commands have the form `// @<commandname> <file> <command args>`

This form was inherited from htmldocck.py. However unlike HTML, the JSON output only produces 1 file. Therefor currently `-` can be used to imply the file for the previous command. Only the first command must specify the file.

It would be nice to automatically infer the name of the relevant file, so `-` is immediately available.

I don't think it makes sense to remove commands having paths, as eventual we probably want to write tests for the interaction between two crates JSON output (see #99513). It would also cause alot of churn.

Doing this also risks making the semantics of `-` more unclear. Is it the previously used file or the default file? These would be equivalent for every statement in the current test suite, but something to keep in mind for the future.

_First discussed by @aDotInTheVoid in https://github.com/rust-lang/rust/pull/100582#discussion_r945829724_

@rustbot modify labels: +A-rustdoc-json +T-rustdoc

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #100588.

r? @aDotInTheVoid

### CODE REVIEW

File: src/tools/jsondocck/src/main.rs [hunk 1]
```diff
@@ -54,27 +54,22 @@ pub enum CommandKind {
 }
 
 impl CommandKind {
-    fn validate(&self, args: &[String], command_num: usize, lineno: usize) -> bool {
+    fn validate(&self, args: &[String], lineno: usize) -> bool {
         let count = match self {
-            CommandKind::Has => (1..=3).contains(&args.len()),
-            CommandKind::Count | CommandKind::Is => 3 == args.len(),
-            CommandKind::Set => 4 == args.len(),
+            CommandKind::Has => (..=2).contains(&args.len()),
```
Comment (aDotInTheVoid) [inline src/tools/jsondocck/src/main.rs:59]: This should be `1..=2`, as the old 0-arg form of `@has` (check file exists) is no longer valid

File: src/tools/jsondocck/src/cache.rs [hunk 2]
```diff
@@ -1,77 +1,50 @@
...
-        if let Some(v) = self.values.get(&path) {
-            return Ok(v.clone());
+    pub fn get_value(&mut self) -> Result<&Value, CkError> {
+        if let Some(ref v) = self.value {
+            return Ok(v);
         }
 
-        let content = self.read_file(path.clone())?;
-        let val = serde_json::from_str::<Value>(&content)?;
+        let val = serde_json::from_str::<Value>(&self.file.content)?;
```
Comment (aDotInTheVoid) [inline src/tools/jsondocck/src/cache.rs:44]: Why do you do this here, instead of deserializing the file content in `Cache::new`?

**Review by aDotInTheVoid — CHANGES_REQUESTED** [review]

**Comment by aDotInTheVoid** [comment]
FYI, I don't have r+ privileges, so can't approve this PR.

r? @CraftSpider

**Comment by GuillaumeGomez** [comment]
I know, but if it looks good to you, I can approve for you, so it's fine. ;)

File: src/tools/jsondocck/src/main.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/tools/jsondocck/src/main.rs:59]: :+1:

File: src/tools/jsondocck/src/cache.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/tools/jsondocck/src/cache.rs:44]: Very good point!

**Comment by GuillaumeGomez** [comment]
Applied your suggestions (and rebased on master to fix CI),

File: src/tools/jsondocck/src/cache.rs [hunk 3]
```diff
@@ -1,77 +1,31 @@
...
-    last_path: Option<PathBuf>,
 }
 
 impl Cache {
     /// Create a new cache, used to read files only once and otherwise store their contents.
-    pub fn new(doc_dir: &str) -> Cache {
+    pub fn new(config: &Config) -> Cache {
+        let root = Path::new(&config.doc_dir);
+        let filename = Path::new(&config.template).file_stem().unwrap();
+        let file_path = root.join(&format!("{}.json", filename.to_str().unwrap()));
```
Comment (aDotInTheVoid) [inline src/tools/jsondocck/src/cache.rs:19]: Nit: Using [`Path::with_extension`](https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.with_extension) will allow running the tool when rust is checked out in a non utf8 path.

File: src/tools/jsondocck/src/cache.rs [hunk 4]
```diff
@@ -1,77 +1,31 @@
...
-            return Ok(v.clone());
-        }
-
-        let content = self.read_file(path.clone())?;
-        let val = serde_json::from_str::<Value>(&content)?;
-
-        self.values.insert(path, val.clone());
-
-        Ok(val)
+    pub fn get_value(&self) -> &Value {
```
Comment (aDotInTheVoid) [inline src/tools/jsondocck/src/cache.rs:28]: Nit: This should be named `pub fn value`, per [C-GETTER](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter), although previously this wasn't followed in jsondocck

**Review by aDotInTheVoid — CHANGES_REQUESTED** [review]

File: src/tools/jsondocck/src/cache.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/tools/jsondocck/src/cache.rs:19]: :o TIL

File: src/tools/jsondocck/src/cache.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/tools/jsondocck/src/cache.rs:28]: :+1:

**Comment by GuillaumeGomez** [comment]
Applied suggestions bis repetita. :)

**Review by aDotInTheVoid — APPROVED** [review]
Looks all good to me.

**Pull request merged**